### PR TITLE
[MERGE WITH MATCHING FRONTEND CHANGES] Feature/ch51 improve pagination performance

### PR DESCRIPTION
**Pull request by nickgravgaard**

### CODE REVIEW

File: service/es_access.py [hunk 1]
```diff
@@ -4,22 +4,31 @@
 
 ELASTICSEARCH_ENDPOINT = app.config['ELASTIC_SEARCH_ENDPOINT']
 MAX_NUMBER_SEARCH_RESULTS = app.config['MAX_NUMBER_SEARCH_RESULTS']
+SEARCH_RESULTS_PER_PAGE = app.config['SEARCH_RESULTS_PER_PAGE']
+
+
+def _get_start_and_end_indexes(page_number):
```
Comment (lgonczar) [inline service/es_access.py]: If we are receiving 1-based page number in the request, it would probably be best to translate it to 0-based page index and use that index everywhere. Not very strong about it, but we may avoid some potential confusion this way.

File: service/server.py [hunk 2]
```diff
@@ -59,32 +60,23 @@ def _hit_postgresql_with_sample_query():
...
-    }
-
-
-def paginated_and_index_address_records(address_records, page_number):
     if address_records:
-        result = paginated_address_records(address_records, page_number)
+        title_numbers = [rec.title_number for rec in address_records]
+        titles = db_access.get_title_registers(title_numbers)
+        ordered = sorted(titles, key=lambda t: title_numbers.index(t.title_number))
+        dicts = [{'title_number': t.title_number, 'data': t.register_data} for t in ordered if t]
```
Comment (lgonczar) [inline service/server.py]: Is the last `if t` needed?

Comment (lgonczar) [inline service/server.py]: How about renaming `dict` to something like `title_data_dicts`?

File: service/server.py [hunk 3]
```diff
@@ -119,18 +111,17 @@ def get_official_copy(title_ref):
 @app.route('/title_search_postcode/<postcode>', methods=['GET'])
 def get_properties(postcode):
```
Comment (lgonczar) [inline service/server.py]: We've got very high level unit tests for this applications. Because of that, we cover a relatively small subset of edge cases. We could use the change this story to improve it and write a few proper tests. Off the top of my head (for both postcode and address search):

- server calls es_utils with the right page number
- server does't call db_utils when es_utils returns no results (and returns an empty page)
- server calls db_utils with the list of title numbers from es_utils
- server returns results from db_utils in the right order

File: service/server.py [hunk 4]
```diff
@@ -119,18 +111,17 @@ def get_official_copy(title_ref):
...
     page_number = int(request.args.get('page', 1))
-    no_underscores = postcode.replace("_", "")
-    no_spaces = no_underscores.replace(" ", "")
-    address_records = es_access.get_properties_for_postcode(no_spaces)
-    result = paginated_and_index_address_records(address_records, page_number)
+    normalised_postcode = postcode.replace('_', '').replace(' ', '')
```
Comment (lgonczar) [inline service/server.py]: :+1:  yes!

**Comment by lgonczar** [comment]
Reviewed. Looks good. Left some suggestions, just to consider. The ones regarding tests are the only important.

File: service/server.py (hunk 3 above)
Comment (nickgravgaard) [inline service/server.py]: This is important, but let's do it in a separate branch

Comment (lgonczar) [inline service/server.py]: :+1:

**Comment by russwillis** [comment]
LGTM
